lightning-graph-sync/src/processing.rs

```diff
@@ -0,0 +1,506 @@
+///
+/// `update_data`: `&[u8]` binary stream that comprises the update data
+pub fn update_network_graph(
+	network_graph: &network_graph::NetworkGraph,
+	update_data: &[u8],
+) -> Result<(), GraphSyncError> {
+	let mut read_cursor = io::Cursor::new(update_data);
+	read_network_graph(&network_graph, &mut read_cursor)
+}
+
+pub(crate) fn read_network_graph<R: Read>(
+	network_graph: &network_graph::NetworkGraph,
+	read_cursor: &mut R,
+) -> Result<(), GraphSyncError> {
+	let mut prefix = Vec::with_capacity(4);
+	prefix.resize(4, 0);
+	read_cursor.read_exact(&mut prefix)?;
+
+	let _deserialize_unsigned = match &prefix[..] {
+		&[76, 68, 75, 1] => true,
+		_ => {
+			return Err(GraphSyncError::ProcessingError(
+				"Prefix must equal 'LDK' and the byte 2".to_string(),
+			));
+		}
+	};
+
+	// let chain_source: Option<Box<dyn lightning::chain::Access>> = None;
+	let current_timestamp = SystemTime::now()
+		.duration_since(UNIX_EPOCH)
+		.unwrap()
+		.as_secs() as u32;
+
+	let retrogressed_timestamp = current_timestamp - TIMESTAMP_RETROGRESSION_DELTA; // back-date updates by one week
+
+	// shadow variable for ownership release after length-restricted reads
+	let mut read_cursor = read_cursor;
+
+	let chain_hash: BlockHash = Readable::read(read_cursor)?;
+
+	let announcement_count: u32 = Readable::read(read_cursor)?;
+	for _ in 0..announcement_count {
+		let features = Readable::read(read_cursor)?;
+		let short_channel_id = Readable::read(read_cursor)?;
+		let node_id_1 = Readable::read(read_cursor)?;
+		let node_id_2 = Readable::read(read_cursor)?;
+
+		let announcement_result = network_graph.update_channel_from_announcement_fragments(
+			short_channel_id,
+			retrogressed_timestamp as u64,
+			features,
+			node_id_1,
+			node_id_2
+		);
+		if let Err(lightning_error) = announcement_result {
+			if let ErrorAction::IgnoreDuplicateGossip = lightning_error.action {
+				// everything is fine, just a duplicate channel announcement
+			} else {
+				return Err(lightning_error.into());
+			}
+		}
+	}
+
+	let update_count: u32 = Readable::read(read_cursor)?;
+	for _current_update_index in 0..update_count {
+		// println!("update index: {}", _current_update_index);
+
+		let short_channel_id: u64 = Readable::read(read_cursor)?;
+		let channel_flags: u8 = Readable::read(read_cursor)?;
+
+		// flags are always sent in full, and hence always need updating
+		let standard_channel_flags = channel_flags & 0b_0000_0011;
+
+		let synthetic_update = if channel_flags & 0b_1000_0000 == 0 {
+			// full update
+			// println!("channel flags: {:b}, scid: {}, full update", channel_flags, short_channel_id);
+
+			let cltv_expiry_delta: u16 = Readable::read(&mut read_cursor)?;
+			let htlc_minimum_msat: u64 = Readable::read(&mut read_cursor)?;
+			let fee_base_msat: u32 = Readable::read(&mut read_cursor)?;
+			let fee_proportional_millionths: u32 = Readable::read(&mut read_cursor)?;
+
+			let has_htlc_maximum_msat: u8 = Readable::read(read_cursor)?;
```

Comment:
Maybe set a flag value instead? ie MAX_UINT64 means "no htlc-max", and the server-side can cap it down to 21 million BTC without loss.

Reply:
I'm not sure this saves space, because if there is no HTLC maximum, sending a u8 takes up less space than a u64.

Reply:
we could do something interesting with BigSize?

Reply:
The sizes are equivalent when the number of present HTLCs == 7 * absent HTLCs. If there are more present HTLCs than absent ones, using the flag saves space, and otherwise, using the u8-prefix saves space. We could add a global u8 flag at the beginning specifying which encoding is used in this transmission, and set it depending on the relation of the set/unset HTLC maxima.

Reply:
My point was based on the assumption that the vast majority of full updates would have a htlc max, but, more importantly, see https://github.com/lightningdevkit/rust-lightning/pull/1155#discussion_r813359975

Reply:
Based on testing, it does indeed appear that the 7:1 ratio of set/updated HTLC maxima vs removed HTLC maxima is preserved, so for now we'll just always use u64::MAX